Added perception module

bitcurious opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->
 
---
 
## Basic Info
 
| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #97 |
| Primary OS tested on | Ubuntu |
| Drone platform tested on | M300 |
 
---
 
## Description of contribution in a few bullet points
 
<!--I added this neat new feature -->
<!-- Also fixed a typo here -->

1. Added perception module.
2. Now one can able to subscribe stereo vision camera stream using this module.
3. Camera parameters for respective stereo cameras is also available.

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
1. Perception was a major module provided by th DJI PSDK but it was not integrated in this wrapper.
2. Stereo / Perperction camera help to perform visual simultaneous localization and mapping (vSLAM).
3. Stereo cameras provide single shot 3D reconstructions of the imaged scene by using perspective changes in two camera views.

 ## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

1. Tested on M300 Drone using OSDK extension module.
2. I tested stereo camera stream and parameteres for respective direction.
3. Start and stop of stream is also tested.
4. Perception module doesn't affect any pre-existed 

## Description of documentation updates required from your changes
 
<!-- Added new parameter, so need to add that to default configs and documentation page -->
<!-- I added some capabilities, need to document them -->
I have updated the `docs/documentation/ThePSDKWrapper.md` file with my changed please add the same chnages on [psdk_ros2 wrapper wiki](https://umdlife.github.io/psdk_ros2/index.html).
 
## Future work that may be required in bullet points
 
<!-- I think there might be some optimizations to be made -->
<!-- I see a lot of redundancy in this package -->

## Screenshots (if appropriate):

vicmassy reviewed (commented):
Hey @bitcurious, thanks a lot for your contribution! That is an awesome piece of work. We will take a look into the code in the upcoming days. In the meantime, if you could fix the linter issues it would be really helpful. You can autoformat based on `.clang-format` file and `CPPLINT.cfg` located at the root of the repository.

umesh-mane commented:
@vicmassy Sure I have solved many of them only 88 character per line exceeeding those are remaining will solve them using autoformat as you mentioned.

umesh-mane commented:
@vicmassy  I have fixed all linter issues please look inti it once you get the time.

biancabnd commented on psdk_wrapper/src/modules/perception.cpp:
```diff
@@ -0,0 +1,404 @@
+    img->encoding = "mono8";
+    img->data = std::vector<uint8_t>(imageRawBuffer, imageRawBuffer + bufferLen);
+    img->header.stamp = this->get_clock()->now();
+    img->header.frame_id = "stereo_cameras_link";
```
you could consider making the frame id a parameter of the perception module so that users could be able to configure it's value, instead of putting it hardcoded

biancabnd reviewed (commented):
@bitcurious very nice development and a great contribution having this new module on the wrapper! :confetti_ball: I left  you some comments and some improvements that can be done. Let me know what you think !

biancabnd commented on psdk_wrapper/src/modules/perception.cpp:
```diff
@@ -0,0 +1,404 @@
+}
+
+// To clear the previous stereo camera stream this function is used.
+void
+PerceptionModule::clear_perception_stereo_cameras_stream()
+{
+  DjiPerception_UnsubscribePerceptionImage(DJI_PERCEPTION_RECTIFY_DOWN);
+  DjiPerception_UnsubscribePerceptionImage(DJI_PERCEPTION_RECTIFY_FRONT);
+  DjiPerception_UnsubscribePerceptionImage(DJI_PERCEPTION_RECTIFY_REAR);
+  DjiPerception_UnsubscribePerceptionImage(DJI_PERCEPTION_RECTIFY_UP);
+  DjiPerception_UnsubscribePerceptionImage(DJI_PERCEPTION_RECTIFY_LEFT);
+  DjiPerception_UnsubscribePerceptionImage(DJI_PERCEPTION_RECTIFY_RIGHT);
+}
```
This method could be optimize. Consider storing the different options in a global variable of some type of vector and use a for loop to unsubscribe from each one of them. Moreover, you could return a boolean indicating the success of the operation

bitcurious commented on docs/documentation/ThePSDKWrapper.md:
```diff
@@ -306,3 +306,17 @@ Moreover, one can check the current storage capacity of the SD card (`psdk_ros2/
 | psdk_ros2/gimbal_angles           | Topic     |
+
+
+### Perception
```
I didn't get which parameters are you refering?
you wnat me to add perception_camera_parameters ?

biancabnd replied:
I was referring to add the new module in the documentation, similar to these: https://github.com/umdlife/psdk_ros2/blob/9858ad443bb1ac038dfa20ec98940931f1d76311/docs/documentation/GettingStarted.md?plain=1#L85-L91

bitcurious replied:
Okay will update the same.

bitcurious commented:
@biancabnd  I have addressed all the comments you raised. Apologies for the multiple commits; I encountered some issues with my GitHub account in VS Code.

biancabnd commented on psdk_wrapper/include/psdk_wrapper/modules/perception.hpp:
```diff
@@ -0,0 +1,175 @@
+  bool deinit();
+
+ private:
+  friend void c_DjiTest_PerceptionImageCallback(T_DjiPerceptionImageInfo imageInfo,
+                                                uint8_t *imageRawBuffer,
+                                                uint32_t bufferLen);
+
+  /* Streaming callbacks */
+  /***
+  * @brief Stereo camera stream of both left and right camera sensor
+  */
+  void DjiTest_PerceptionImageCallback(T_DjiPerceptionImageInfo imageInfo,
+                                       uint8_t *imageRawBuffer,
+                                       uint32_t bufferLen);
```
@bitcurious can you please address this as well? Also, make sure to use snake case format in all the methods names as that is the standard used across the entire wrapper. Thank you.

bitcurious replied:
I've removed the DJITest tag and applied clang-format to update the files. I apologize for the multiple commits caused by the issue that prevented the updates from being applied smoothly.

https://github.com/bitcurious/psdk_ros2/blob/301f9f00781c03b1172e21ee360e115130f51824/psdk_wrapper/include/psdk_wrapper/modules/perception.hpp#L110-L119

biancabnd replied:
Hi @bitcurious, no problem!
However, what I meant with snake_case is to rename the methods something like this:  c_perception_image_callback. So we use snake case format  and not camel case. Thanks

bitcurious replied:
@biancabnd Understood, I will do updated commit.

biancabnd reviewed (approved):
RTBM! :confetti_ball:
